src/datanode/src/heartbeat/task_tracker.rs

```diff
@@ -0,0 +1,299 @@
+// Copyright 2023 Greptime Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::HashMap;
+use std::sync::Arc;
+use std::time::Duration;
+
+use futures_util::future::BoxFuture;
+use snafu::ResultExt;
+use store_api::storage::RegionId;
+use tokio::sync::watch::{self, Receiver};
+use tokio::sync::RwLock;
+use tokio::{select, time};
+
+use crate::error::{self, Error, Result};
+
+/// The state of a async task.
+#[derive(Debug, Default, Clone)]
+pub(crate) enum TaskState<T: Send + Sync + Clone> {
+    Error(Arc<Error>),
+    #[default]
+    Running,
+    Done(T),
+}
+
+pub(crate) type TaskWatcher<T> = Receiver<TaskState<T>>;
+
+pub async fn wait<T: Send + Sync + Clone>(watcher: &mut TaskWatcher<T>) -> Result<T> {
+    loop {
+        watcher
+            .changed()
+            .await
+            .context(error::WatchAsyncTaskChangeSnafu)?;
+
+        let r = &*watcher.borrow();
+        match r {
+            TaskState::Error(err) => return Err(err.clone()).context(error::AsyncTaskExecuteSnafu),
+            TaskState::Running => {}
+            TaskState::Done(value) => return Ok(value.clone()),
```

Comment:
Just wondering if it's possible to avoid clone here, considering the watcher is mut borrowed. Maybe Done(T) can be Done(Option<T>) and then take the value.

Reply:
In our use case, it only clones once(the state machine is naive; transitions: running -> error or running -> done)